packages/cubejs-questdb-driver/src/QuestDriver.ts

```diff
@@ -0,0 +1,225 @@
+      database: process.env.CUBEJS_DB_NAME,
+      port: <any>process.env.CUBEJS_DB_PORT,
+      user: process.env.CUBEJS_DB_USER,
+      password: process.env.CUBEJS_DB_PASS,
+      ...config
+    });
+    this.pool.on('error', (err) => {
+      console.log(`Unexpected error on idle client: ${err.stack || err}`);
+    });
+
+    this.config = {
+      ...this.getInitialConfiguration(),
+      ...config,
+    };
+  }
+
+  protected getInitialConfiguration(): Partial<QuestDriverConfiguration> {
+    return {
+      readOnly: true,
+    };
+  }
+
+  protected getTypeParser = (dataTypeID: TypeId, format: TypeFormat | undefined) => {
+    const isTimestamp = timestampDataTypes.includes(dataTypeID);
+    if (isTimestamp) {
+      return timestampTypeParser;
+    }
+
+    const parser = types.getTypeParser(dataTypeID, format);
+    return (val: any) => parser(val);
+  };
+
+  protected getQuestTypeForField(dataTypeID: number): string | null {
+    if (dataTypeID in NativeTypeToQuestType) {
+      return NativeTypeToQuestType[dataTypeID].toLowerCase();
+    }
+
+    return null;
+  }
+
+  public async testConnection(): Promise<void> {
+    await this.pool.query('SELECT $1 AS number', ['1']);
+  }
+
+  protected mapFields(fields: FieldDef[]) {
+    return fields.map((f) => {
+      const questType = this.getQuestTypeForField(f.dataTypeID);
+      if (!questType) {
+        throw new Error(
+          `Unable to detect type for field "${f.name}" with dataTypeID: ${f.dataTypeID}`
+        );
+      }
+
+      return ({
+        name: f.name,
+        type: this.toGenericType(questType)
+      });
+    });
+  }
+
+  protected async queryResponse(query: string, values: unknown[]) {
+    const conn = await this.pool.connect();
+
+    try {
+      const res = await conn.query({
+        text: query,
+        values: values || [],
+        types: {
+          getTypeParser: this.getTypeParser,
+        },
+      });
+      return res;
+    } finally {
+      conn.release();
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public async query<R = unknown>(query: string, values: unknown[], options?: QueryOptions): Promise<R[]> {
+    const result = await this.queryResponse(query, values);
+    return result.rows;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public async downloadQueryResults(query: string, values: unknown[], options: DownloadQueryResultsOptions) {
+    const res = await this.queryResponse(query, values);
+    return {
+      rows: res.rows,
+      types: this.mapFields(res.fields),
+    };
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public async createSchemaIfNotExists(schemaName: string): Promise<any> {
+    // no-op as there are no schemas in QuestDB
+  }
+
+  public async tablesSchema() {
+    const tables = await this.getTablesQuery('');
+
+    const metadata: Record<string, Record<string, object>> = { '': {} };
```

Comment:
The thing is that QuestDB doesn't have a notion of schema/logical database while this driver method has to return a `{ 'schema_name': { 'table1': {...} } }` object. I tried to use a non-empty hardcoded schema name here (say, 'global'), but it leads to `global.table_name` being used in the generated queries. So, using an empty string here is a mere hack to work around this problem.

If there is already a way to solve it in a proper way, I'm all for it. Otherwise, I could maybe add a `supportsSchema(): boolean` driver method and try to use it in cube-cli and/or dev server.

WDYT?

Reply:
I see. Does '' work, or do you get `.table_name`? If you could add a comment in the source explaining what you already explained above, that would suffice.

Reply:
Makes sense. Added a comment in 6ca4b88d350e5de897eb4c9aac23fec896fcd923